host/docs/shd_image_loader.1
.TH "shd_image_loader" 1 "3.9.0" SHD "User Commands"
.SH NAME
shd_image_loader - SHD Image Loader

.SH DESCRIPTION
Burn firmware and FPGA images onto connected Ettus Research devices.

.SH SYNOPSIS
.B  shd_image_loader [OPTIONS]

.SH OPTIONS
.IP "List options:"
--help
.IP "Device and loader arguments:"
--args=""
.IP "Custom firmware filepath:"
--fw-path=""
.IP "Custom FPGA filepath:"
--fpga-path=""
.IP "Don't burn firmware:"
--no-fw
.IP "Don't burn FPGA:"
--no-fpga

.SH SPECIFYING A PARTICULAR DEVICE
.sp
Without any arguments given through the \fIargs=""\fR option, this utility will start a session
with the first Ettus Research device it sees. The arguments shown below will narrow down the
utility's search for a particular device:

.SS All devices
.sp
The \fIname=\fR and \fIserial=\fR arguments can be used to specify any device except an
uninitialized USB device or an OctoClock bootloader. The \fItype=\fR argument can be used
to specify any device, as described below:

"type=" argument

Argument       |     Device

type=smini2     |     SMINI N200, SMINI N210

type=b200      |     SMINI B200, SMINI B210

type=e100      |     SMINI E100, SMINI E110

type=e3x0      |     SMINI E310

type=octoclock |     OctoClock

.sp
NOTE: The SMINI1, SMINI2, and SMINI B100 are not supported.

.SS Network devices
.sp
By default, this utility will check all network interfaces for network-based devices, but a specific
IP address can be specified with the \fIaddr=\fR argument.

.SS NI-RIO devices
.sp
An X-Series devices connected via PCIe can be specified by its NI-RIO resource through the \fIresource=\fR
argument, and the RPC port through which to communicate with it can be specified with the \fIrpc-port=\fR
argument. Using these options is not recommended, as their default values are almost always used.

.SS OctoClock devices
An OctoClock's name and serial are only exposed when the firmware is loaded, so if the device only has a
bootloader, only the \fIaddr=\fR argument can be used to find it.

.SH DEVICE-SPECIFIC LOADER OPTIONS
.sp
Certain devices have specific options for customizing their image loading process, and these can be passed
in through the \fI--args=""\fR option. These arguments are specified below:

.SS SMINI N200, SMINI N210
.sp
The \fIoverwrite-safe\fR option will overwrite the device's safe-mode firmware and FPGA images. This is
NOT RECOMMENDED, as these images serve as the backup if the device's primary images are corrupted.

.sp
The \fIreset\fR option will automatically reset the device once the loading process finishes. When the
device resets, it will have its new images loaded.

.SS SMINI X300, SMINI X310 (Ethernet only)
.sp
The \fIconfigure\fR option will automatically reset the device once the loading process finishes. When
the device resets, it will have its new FPGA image loaded.

.sp
The \fIverify\fR option will tell the device to internally verify the integrity of the image as it loads.
This greatly increases the loading time.

.SH EXAMPLES

.SS Load only the default FPGA image onto a specific N2x0 device and reset
.sp
shd_image_loader --args="type=smini2,addr=192.168.10.2,reset" --no-fw
.ft

.SS Load a custom FPGA image onto a specific X3x0 device
.sp
shd_image_loader --args="type=x300,addr=192.168.40.2" --fpga-path="/home/user/my_x300_fpga_image.bit"
.ft

.fi

.SH SEE ALSO
SHD documentation:
.B http://files.ettus.com/manual/
.LP
GR-SHD documentation:
.B http://gnuradio.org/doc/doxygen/page_shd.html
.LP
Other SHD programs:
.sp
shd_images_downloader(1) smini2_card_burner(1)
.SH AUTHOR
This manual page was written by Nicholas Corgan
for the Debian project (but may be used by others).
.SH COPYRIGHT
Copyright (c) 2015 Ettus Research LLC
.LP
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.
.LP
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.


host/utils/shd_image_loader.cpp
//
// Copyright 2015 Ettus Research LLC
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

#include <csignal>
#include <cstdlib>
#include <iostream>

#include <boost/assign.hpp>
#include <boost/filesystem.hpp>
#include <boost/format.hpp>

#include <shd/config.hpp>
#include <shd/image_loader.hpp>
#include <shd/types/device_addr.hpp>
#include <shd/utils/safe_main.hpp>
#include <boost/program_options.hpp>

namespace fs = boost::filesystem;
namespace po = boost::program_options;

static std::string device_type = "";
static int num_ctrl_c = 0;

/*
 * If the user presses Ctrl+C, warn them that they may corrupt their device.
 * If they press it again, provide instructions on restoring the device
 * (if applicable) and exit.
 */
void sigint_handler(int){
    num_ctrl_c++;
    if(num_ctrl_c == 1){
        std::cout << std::endl
                  << "Are you sure you want to abort? If you do, your device will likely" << std::endl
                  << "be in an unstable or unusable state." << std::endl
                  << "Press Ctrl+C again to abort." << std::endl << std::endl;
    }
    else{
        std::cout << std::endl << shd::image_loader::get_recovery_instructions(device_type) << std::endl;
        exit(EXIT_FAILURE);
    }
}

int SHD_SAFE_MAIN(int argc, char *argv[]){

    std::string fw_path = "";
    std::string fpga_path = "";

    po::options_description desc("Allowed options");
    desc.add_options()
        ("help", "help message")
        ("args", po::value<std::string>()->default_value(""), "Device args, optional loader args")
        ("fw-path", po::value<std::string>(&fw_path)->default_value(""), "Firmware path (uses default if none specified)")
        ("fpga-path", po::value<std::string>(&fpga_path)->default_value(""), "FPGA path (uses default if none specified)")
        ("no-fw", "Don't burn firmware")
        ("no-fpga", "Don't burn FPGA")
    ;

    po::variables_map vm;
    po::store(po::parse_command_line(argc, argv, desc), vm);
    po::notify(vm);

    // Help message
    if (vm.count("help")){
        std::cout << "SHD Image Loader" << std::endl
                  << std::endl
                  << "Load firmware and/or FPGA images onto an Ettus Research device." << std::endl
                  << std::endl
                  << desc << std::endl;
        return EXIT_FAILURE;
    }

    // Convert user options
    shd::image_loader::image_loader_args_t image_loader_args;
    image_loader_args.args          = vm["args"].as<std::string>();
    image_loader_args.load_firmware = (vm.count("no-fw") == 0);
    image_loader_args.load_fpga     = (vm.count("no-fpga") == 0);
    image_loader_args.firmware_path = vm["fw-path"].as<std::string>();
    image_loader_args.fpga_path     = vm["fpga-path"].as<std::string>();

    // Force user to specify a device
    if(not image_loader_args.args.has_key("type")){
        throw shd::runtime_error("You must specify a device type.");
    }

    // Clean up paths, if given
    if(image_loader_args.firmware_path != ""){
        #ifndef SHD_PLATFORM_WIN32
        if(image_loader_args.firmware_path.find("~") == 0){
            image_loader_args.firmware_path.replace(0,1,getenv("HOME"));
        }
        #endif /* SHD_PLATFORM_WIN32 */
        image_loader_args.firmware_path = fs::absolute(image_loader_args.firmware_path).string();
    }
    if(image_loader_args.fpga_path != ""){
        #ifndef SHD_PLATFORM_WIN32
        if(image_loader_args.fpga_path.find("~") == 0){
            image_loader_args.fpga_path.replace(0,1,getenv("HOME"));
        }
        #endif /* SHD_PLATFORM_WIN32 */
        image_loader_args.fpga_path = fs::absolute(image_loader_args.fpga_path).string();
    }

    // Detect which type of device we're working with
    device_type = image_loader_args.args.get("type","");

    std::signal(SIGINT, &sigint_handler);
    if(not shd::image_loader::load(image_loader_args)){
        std::cerr << "No applicable SHD devices found" << std::endl;
        return EXIT_FAILURE;
    }

    return EXIT_SUCCESS;
}
